docs/tpm-simulators.md

```diff
@@ -0,0 +1,162 @@
+
+It is possible to use TPM simulators instead of real TPMs **for testing purposes
+only**. Two simulators were tested and are known to be working with Fobnail:
+[Microsoft's](https://github.com/microsoft/ms-tpm-20-ref) and
+[IBM's](https://sourceforge.net/projects/ibmswtpm2/). Refer to their respective
+documentations for build and launch instructions.
+
+## Differences between simulated and real TPM
+
+The main difference is that simulator is started after system is up. This means
+that there will be no measurements of firmware nor OS. Attestation is based on
+these measurements, without them there is nothing to attest to, which makes use
+of simulated TPM useless for practical applications. It also means that nothing
+called `TPM2_Startup` command. Attester is able to detect this and calls that
+function automatically.
+
+Hardware and firmware TPM vendors give assurance of proper functioning of their
+TPMs. This is done by creating a certificate for each TPM separately,
+specifically for their Endorsement Keys (EKs). These certificates are created
+during TPM manufacturing and provisioning process, and are written to NVRAM.
+They point to their signing certificates through `authorityInfoAccess` X509V3
+extension. The chain continues until a self-signed root that is implicitly
+trusted. That root must be known by Fobnail Token in advance in order to trust
+given TPM.
+
+For simulated TPM, such certificate has to be manually created and injected into
+NVRAM. EK certificate is created for EK created from EPS (Endorsement Primary
+Seed), which is randomly created when the simulator is started for the first
+time. Nobody is issuing certificates for all instances of simulated TPMs. One of
+the reasons is probably the fact that emulators are not fully compatible with
+specification, especially when it comes to hardware protection mechanisms
+(Physical Presence, protected storage etc.).
+
+There are also other small differences, e.g. after non-orderly shutdown `safe`
+field in `TPMS_CLOCK_INFO` structure is set after 12 seconds in simulator, while
+specification allows for up to 2^22 milliseconds (around 70 minutes) to help
+with NVRAM wear leveling. Note that this field is checked by Fobnail during
+attestation.
+
+## Provisioning simulated TPM
+
+As mentioned, EK certificate must be written to TPM NVRAM. In addition, it must
+point to valid CA certificate that will be downloaded during platform
+provisioning. A script for simplifying this process and required configuration
+are included in [Attester's repository](https://github.com/fobnail/fobnail-attester/tree/main/tools).
+
+Configuration assumes that CA certificate will be made available at address
+`http://127.0.0.1:8080/ca_cert.der`, to change this please edit `ek_v3.ext` in
+`tools` directory.
+
+To create root certificate, EK certificate and write the latter to NVRAM, it is
+enough to call the following (`-s` tells to send `TPM2_Startup` command):
+
+```
+$ ./tools/tpm_manufacture.sh -s
```

Comment:
```
[23:04:10] pietrushnic:fobnail-attester git:(main) $ ./tools/tpm_manufacture.sh -s
Sending TPM2_Startup command
'flushcontext -t' failed, aborting.
```

Reply:
This is on the other side of the keyboard:

```
[23:17:56] pietrushnic:fobnail-attester git:(main) $ cat /tmp/tpm_manufacture.log 
Starting log file at: Wed 26 Apr 2023 11:17:56 PM CEST
./tools/tpm_manufacture.sh: line 42: tpm2_startup: command not found
./tools/tpm_manufacture.sh: line 26: tpm2_flushcontext: command not found
[23:18:14] pietrushnic:fobnail-attester git:(main) $ tpm2 -v
zsh: command not found: tpm2
```

But do we have any instructions on how to install the TPM simulator? Also, maybe this is not part of the build instructions cleanup, and I should not review that at this point.

Reply:
So no `tpm2-tools` installed at all... I added requirements in https://github.com/fobnail/docs/pull/47/commits/24c61127cf5e79c93abf14ba615b5a7e7567cb5b

> But do we have any instructions on how to install the TPM simulator?

First paragraph redirects to simulator's instructions.

Reply:
This helps only partially because you have to install simulated TPM. It could be more straightforward. It has to be done. Choosing between two and not specifying the version will likely lead to questions and issues. Also, why didn't you consider [swtpm](https://github.com/stefanberger/swtpm), which seems to be a state of an art TPM simulator?

Also, it could be clearer how to use the MSFT TPM simulator, but luckily I manage to get through that. It could be more straightforward. Maybe let's note that for future improvements and close this conversation?

Reply:
I added build and run instructions in https://github.com/fobnail/docs/pull/47/commits/adcc25cc3e3552fe58afa432e140119b3d9ef218

As I said in one of the other comments, I don't have any experience with swtpm. It wasn't tested. Contribution welcome :slightly_smiling_face:

Reply:
https://github.com/fobnail/docs/issues/56